rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/CallHierarchy.java

```diff
@@ -0,0 +1,138 @@
+            TF.sourceLocationType(), CallHierarchyFields.DEFINITION,
+            TF.sourceLocationType(), CallHierarchyFields.SELECTION
+        );
+    }
+
+    public IConstructor direction(Direction dir) {
+        switch (dir) {
+            case INCOMING: return this.incoming;
+            case OUTGOING: return this.outgoing;
+            default: throw new IllegalArgumentException();
+        }
+    }
+
+    public CallHierarchyItem toLSP(IConstructor cons, ColumnMaps columns) {
+        var name = ((IString) cons.get(CallHierarchyFields.NAME)).getValue();
+        var kind = DocumentSymbols.symbolKindToLSP((IConstructor) cons.get(CallHierarchyFields.KIND));
+        var def = (ISourceLocation) cons.get(CallHierarchyFields.DEFINITION);
+        var definitionRange = Locations.toRange(def, columns);
+        var selection = (ISourceLocation) cons.get(CallHierarchyFields.SELECTION);
+        var selectionRange = Locations.toRange(selection, columns);
+
+        var ci = new CallHierarchyItem(name, kind, def.top().getURI().toString(), definitionRange, selectionRange);
+        var kws = cons.asWithKeywordParameters();
+        if (kws.hasParameter(CallHierarchyFields.TAGS)) {
+            ci.setTags(DocumentSymbols.symbolTagsToLSP((ISet) kws.getParameter(CallHierarchyFields.TAGS)));
+        }
+        if (kws.hasParameter(CallHierarchyFields.DETAIL)) {
+            ci.setDetail(((IString) kws.getParameter(CallHierarchyFields.DETAIL)).getValue());
+        }
+        if (kws.hasParameter(CallHierarchyFields.DATA)) {
+            ci.setData(serializeData((IConstructor) kws.getParameter(CallHierarchyFields.DATA)));
+        }
+
+        return ci;
+    }
+
+    /**
+     * Serializes the call hierarchy item data field as a {@link String}.
+     * The deserialization counterpart is {@link org.rascalmpl.vscode.lsp.parametric.ILanguageContributions#parseCallHierarchyData}.
+     */
+    private String serializeData(IConstructor data) {
+        return data.toString();
+    }
+
+    public CompletableFuture<IConstructor> toRascal(CallHierarchyItem ci, Function<String, CompletableFuture<IConstructor>> dataParser, ColumnMaps columns) {
+        var serializedData = ci.getData() != null
+            ? ((JsonPrimitive) ci.getData()).getAsString()
+            : "";
+        return dataParser.apply(serializedData).thenApply(data ->
+            VF.constructor(callHierarchyItemCons, List.of(
+                VF.string(ci.getName()),
+                DocumentSymbols.symbolKindToRascal(ci.getKind()),
+                Locations.setRange(Locations.toLoc(ci.getUri()), ci.getRange(), columns),
+                Locations.setRange(Locations.toLoc(ci.getUri()), ci.getSelectionRange(), columns)
+            ).toArray(new IValue[0]), Map.of(
+                CallHierarchyFields.TAGS, DocumentSymbols.symbolTagsToRascal(ci.getTags()),
+                CallHierarchyFields.DETAIL, VF.string(ci.getDetail()),
```

Comment:
both `getTags()` and `getDetail` can return null, this code is not safe for that. We should only set the entries in the map if they're actually set. Assuming the defaults in rascal are properly modelled.